### PR TITLE
Switch to the I-beam cursor when hovering over the terminal

### LINKED ISSUE #1441: Mouse pointer should be text and not arrow one
# Expected behavior
The mouse pointer used should be the text cursor (Ꮖ) one since text is selectable, not the default arrow (↖) pointer.

# Actual behavior
It's the default arrow pointer that's used.

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
This commit makes us use the I-beam cursor when the user hovers over the
terminal, *unless* mouse mode is enabled. I've also plumbed up a bunch
of events so that:

* If mouse mode is _toggled_ while hovering, the cursor will switch to
  the arrow if it's on or the I-beam if it's off.
* If you hold down shift to suppress mouse mode, the cursor will switch
  back to the I-beam.

Fixes #1441.

## Detailed Description of the Pull Request / Additional comments
It turns out that we're in violation of the Windows human interface guidelines, so that's good.

It also turns out that _you cannot set the cursor type of a control_ -- you actually have to _actively set and reset the cursor_ when the pointer enters or exits.

terminal_mouse
_my selection color is bright red to differentiate it from vim's mouse mode selection_

### CODE REVIEW

**Comment by zadjii-msft** [comment]
On a scale of yes to no, could this be made a setting (in a future PR) so I could disable this? Should it be a global or a profile setting?

**Comment by DHowett-MSFT** [comment]
In a future PR, it could probably be made a setting -- global, though.

**Comment by mdtauk** [comment]
If it is implemented, it should work with the new Beam cursor options.

image

**Comment by DHowett-MSFT** [comment]
@mdtauk that is not an i-beam mouse cursor. whatever that is works with Terminal as of v0.10 (or v0.11? which is not yet out; it merged recently.)

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 1]
```diff
@@ -766,6 +779,23 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
         e.Handled(handled);
     }
 
+    void TermControl::_KeyUpHandler(winrt::Windows::Foundation::IInspectable const& /*sender*/,
+                                    Input::KeyRoutedEventArgs const& e)
+    {
+        // If the current focused element is a child element of searchbox,
+        // we do not send this event up to terminal
+        if (_searchBox && _searchBox->ContainsFocus())
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControl.cpp]: Presumably this bit would go away with #4748?

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 2]
```diff
@@ -889,6 +919,18 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
         return _terminal->IsTrackingMouseInput();
     }
 
+    // Method Description:
+    // - Handles changes in mouse mode state
+    winrt::fire_and_forget TermControl::_TerminalMouseModeChanged()
+    {
+        co_await Dispatcher();
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControl.cpp]: Didn't think that would work without a param, that's neat.

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 3]
```diff
@@ -1150,6 +1192,45 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+
+        auto coreWindow = Window::Current().CoreWindow();
+        _oldCursor = coreWindow.PointerCursor();
+
+        if (_terminal->IsTrackingMouseInput())
+        {
+            return;
+        }
+
+        coreWindow.PointerCursor(_textCursor);
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControl.cpp]: What if we've got two SxS panes, press shift in one, then hover over the other?

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 4]
```diff
@@ -889,6 +919,18 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
 
+    // Method Description:
+    // - Handles changes in mouse mode state
+    winrt::fire_and_forget TermControl::_TerminalMouseModeChanged()
+    {
+        co_await Dispatcher();
+        if (_oldCursor) // if we have an active cursor transition
+        {
+            auto coreWindow = Window::Current().CoreWindow();
+            coreWindow.PointerCursor(_CanSendVTMouseInput() ? _pointerCursor : _textCursor);
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControl.cpp]: Should we have a helper method that's just "set the `PointerCursor` to the correct cursor for our current mouse input mode"?

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 3 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalControl/TermControl.cpp]: That works fine. The cursor is updated when the pointer enters and leaves the control.

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 2 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalControl/TermControl.cpp]: yeah! It's supes cool

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 1 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalControl/TermControl.cpp]: Yep!

**Comment by DHowett-MSFT** [comment]
So, I think this is your setting:

image

Cheekiness aside, I'm not averse to speccing a global setting that configures how the cursor works -- but it _will be_ a case of users opting for less expressiveness/fewer features, so we should almost certainly measure its use.

**Comment by DHowett-MSFT** [comment]
This is in selfhost. Give it a feel.

**Comment by zadjii-msft** [comment]
* [ ] The cursor remains an I-beam when I mouse over the tabs
* [ ] If I open tmux (with mouse mode) in one tab, then hover another tab, the cursor switches back to the I-beam, and stays in an I beam even if I click inside tmux

**Comment by zadjii-msft** [comment]
showerthought: what if we just always turned the cursor back into an arrow, then let the new control change it to whatever they want? Presumably one control's `PointerExited` is fired _before_ the next control's `PointerEntered`, right? Or does that cause flickering?

Additionally: Are we planning on putting this is 1.0?

**Comment by DHowett-MSFT** [comment]
not about about 1.0. My concern with always setting Arrow is that what if they're hovering with a pen? I dunno, I kinda wanna ask AustinL.

**Comment by bropines** [comment]
Sorry Comrades for what I write in this branch (I haven't figured out yet with GitHub), but I have a bug because of which this program does not allow me to write anything.

**Comment by DHowett-MSFT** [comment]
@bropines it looks like you are running into #4448

**Comment by egmontkob** [comment]
So this is about the mouse pointer (mouse cursor), and not the text insertion cursor. I legit had to read the thread halfway through to realize it.

Don't you guys always confuse these two? I do. Similar names, the word "cursor" overloaded for them, and both possibly being of "I-beam" shape.

My preference is to say "mouse pointer" for, well, that one; without using the word "cursor". Do you guys have a preferred terminology?

**Comment by bropines** [comment]
> @bropines похоже, что вы бежите в #4448

Thank you very much. I wouldn't have guessed it myself

**Comment by DHowett-MSFT** [comment]
@egmontkob yep, this is about the mouse pointer. I should endeavour to be better about saying "pointer" when I mean the thing the user points with. :smile:

Alas, overloaded terms are pervasive in our industry. They're so hard to escape.

**Comment by DHowett-MSFT** [comment]
(And yeah, we get them mixed up all the time.)

File: src/cascadia/TerminalControl/TermControl.h [hunk 5]
```diff
@@ -167,6 +167,10 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
 
         winrt::Windows::UI::Xaml::Controls::SwapChainPanel::LayoutUpdated_revoker _layoutUpdatedRevoker;
 
+        std::optional<winrt::Windows::UI::Core::CoreCursor> _oldCursor; // when we toggle the cursor, we have to save it here to restore it
+        winrt::Windows::UI::Core::CoreCursor _textCursor;
+        winrt::Windows::UI::Core::CoreCursor _pointerCursor;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControl.h:172]: (Correct me if I'm wrong) It looks like you only set these in the constructor. Why not just make them `constexpr` in the namespace?

Comment (DHowett) [reply src/cascadia/TerminalControl/TermControl.h:172]: they're winrt objects. it sucks.

**Comment by jsoref** [comment]
The conflicts shouldn't be hard to resolve `whitelist` was renamed to `expect` -- If you need help, I can help on Sunday.

**Comment by angelog0** [comment]
All the decent Terminals I know have this behavior: when the mouse pointer is over them it changes into a sort of `I`. This makes selecting text more intuitive and practical...

**Comment by DHowett** [comment]
And you are commenting on this open pull request that does exactly this so as to move it along?

**Comment by dufferzafar** [comment]
Is this still being worked on?

**Comment by DHowett** [comment]
Not actively.

**Comment by miniksa** [comment]
Closing as it's super stale, but not deleting the branch in case someone wants to dig it up again one day.

**Comment by zadjii-msft** [comment]
To add some notes that never made it here a few years back: 

I _believe_ there was no good way for us to set the mouse pointer for a single control in XAML. A control could change the cursor in PointerEntered, but IIRC there's a race between PointerEnter's and PointerExits. So when you've got an app that's got all sorts of mixed UI elements, some that want the cursor to be an I-beam and some that want the arrow, well now you're stuck. As you hover over different elements, they'll fight over who's actually setting the cursor to what.

**Pull request closed without merging**